Add a riak plugin

jchristgit opened the pull request:
<!--
Describe the change in summary section, including rationale and degin decisions.
Include "Fixes #nnn" if you are fixing an existing issue.

In "Component Name" section write which component is changed in this PR. This
will help us review your PR quicker.

If you have more information you want to add, write them in "Additional
Information" section. This is usually used to help others understand your
motivation behind this change. A step-by-step reproduction of the problem is
helpful if there is no related issue.
-->

##### Summary

Fixes: #2413

Add a plugin for monitoring Riak KV instances.

##### Component Name

collectors/python.d.plugin/riak

##### Additional Information

As discussed in #2413, this adds a plugin for monitoring Riak KV instances. The plugin uses Riak's /stats endpoint to determine statistics about it.

Healthchecks are also included, but I did not get them to display in the Alarms page on the web dashboard, which leads me to think that they might not work as expected. I dropped the configuration file into `/etc/netdata/health.d/riak.conf` but it seems there's something going wrong there. Would be happy for any pointers.

As far as the [Pull Request Checklist](https://docs.netdata.cloud/collectors/python.d.plugin/#pull-request-checklist-for-python-plugins) goes, it seems some paths over there are changed, so I hope I found the correct configuration files :smile:

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/netdata/netdata?pullRequest=6286) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/netdata/netdata?pullRequest=6286) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/netdata/netdata?pullRequest=6286) it.</sub>

ilyam8 commented on collectors/python.d.plugin/python.d.conf:
```diff
@@ -88,6 +88,7 @@ nginx_log: no
 # redis: yes
 # rethinkdbs: yes
 # retroshare: yes
+# riak: yes
```
i think we need to name it `riakkv` or `riakkv`

jchristgit replied:
Renamed in d7f04df1db2cb0967b8482fa0a4ecf6a50eb1795.

ilyam8 commented on collectors/python.d.plugin/riak/riak.chart.py:
```diff
@@ -0,0 +1,528 @@
+                self.keys.append(m + "_" + k)
+        for k in ["min", "max", "mean", "median", "total"]:
+            for m in VNODEQ_KEYS:
+                self.keys.append(m + "_" + k)
```
I suggest to move keys building from init. For example, you could make it a simple function

```python

def keys():
    ...

....
....

self.keys = keys()

```

jchristgit replied:
Custom keys building has been removed altogether in 7469f4d2d13abd74c1b5c9a0ba1d9eebc43f7496 per my comment below.

ilyam8 commented on collectors/python.d.plugin/riak/riak.chart.py:
```diff
@@ -0,0 +1,528 @@
+            return None
+
+        data = dict()
+        stats = loads(raw)
```
use `try, except` for loads `(TypeError, ValueError)`

jchristgit replied:
Wrapped in `try` / `except` in fa4e72e18b237a1333f8c2aad2e66762eb47940d. Is there anything else I should do in case of error besides `return None`?

ilyam8 replied:
no :+1:

ilyam8 commented:
@jchristgit very nice :+1: 	And please sign the CLA

ilyam8 commented on collectors/python.d.plugin/riak/riak.chart.py:
```diff
@@ -0,0 +1,528 @@
+# default module values (can be overridden per job in `config`)
+update_every = 2
+priority = 60000
+retries = 60
```
`priority = 60000` and `retries = 60` are defaults, remove them

> update_every = 2

please add a comment why default update every is 2, and add it to the README to

jchristgit replied:
Documented in ba3df872d7357a877a26b25b669bcd8e04add5c6 with defaults removed.

jchristgit commented:
@ilyam8 Thanks for the feedback! I'll apply and test your suggestions on Monday to make sure I don't coincidentally break something :D

A question regarding the CLA - from my understanding the CLA suggests that I wrote all of the code myself, but the original base implementation was written by you (see https://github.com/netdata/netdata/issues/2413#issuecomment-312923018). Should I add a comment in the code for attribution or how does that work?

jchristgit commented:
After some more thinking, could we get rid of the manual `if x in keys` when loading the response altogether? It would remove a lot of code that we currently have for figuring out which keys to parse in the response (and a lot of them are unused currently already)

ilyam8 commented:
> Should I add a comment in the code for attribution or how does that work?

Nah


> After some more thinking, could we get rid of the manual if x in keys when loading the response altogether? I

yes, just return `json.loads` result

ilyam8 commented on collectors/python.d.plugin/python.d.conf:
```diff
@@ -88,6 +88,7 @@ nginx_log: no
 # redis: yes
 # rethinkdbs: yes
 # retroshare: yes
+# riak: yes
```
riak_kv

jchristgit replied:
Missed that! Fixed in fe218783995f039712fc4feca8b446a331c64762.

ilyam8 commented on web/gui/dashboard.js:
```diff
@@ -4194,23 +4194,6 @@ NETDATA.peityChartCreate = function (state, data) {
-    return true;
-}
-
-NETDATA.textOnlyUpdate = NETDATA.textOnlyCreate;
```
why these changes?

jchristgit replied:
Honestly, no clue. Reset to master in 03ba0d113bcc72f6b0e3f3e56625a6ee1a6501e8.

ilyam8 commented on web/gui/dashboard_info.js:
```diff
@@ -267,6 +267,12 @@ netdataDashboard.menu = {
         info: 'Performance metrics for <b>RetroShare</b>. RetroShare is open source software for encrypted filesharing, serverless email, instant messaging, online chat, and BBS, based on a friend-to-friend network built on GNU Privacy Guard (GPG).'
     },
 
+    'riak': {
```
`riak_kv`

@vlvkobal 	please remind me  - do we have any problems with `_` underscore in names ? i recall you faced some issues

jchristgit replied:
@ilyam8 Doesn't this apply to the metrics though? The metrics are namespaced under `riak.` currently (e.g. `riak.kv`, `riak.dt` and so on). I tested this with `riak_kv` earlier but I recall the information here wasn't applied anymore

ilyam8 replied:
> I tested this with riak_kv earlier but I recall the information here wasn't applied anymore

That is why i summon @vlvkobal 	:smiling_imp:

ilyam8 replied:
Thanks. So we need to remove `_` from the name. How about to name it simply `riakkv` @jchristgit?

jchristgit replied:
Yup, renamed in bfcaf640dbfd7ac7afba8bcf98f26bed110803b5.

ilyam8 commented on web/gui/dashboard_info.js:
```diff
@@ -267,6 +267,12 @@ netdataDashboard.menu = {
     },
 
+    'riak': {
+        title: 'Riak',
```
Riak KV

ilyam8 replied:
> title

jchristgit replied:
Changed in 71baea7939079c01b75878ac094681a5c093e16d.

ilyam8 commented on collectors/python.d.plugin/riak_kv/riak_kv.chart.py:
```diff
@@ -0,0 +1,314 @@
+
+        try:
+            return loads(raw)
+        except (TypeError, ValueError):
```
oh, i forgot, my bad

please log the error if any

jchristgit replied:
Added in 49e8aee8d9ed99808a77c44b7bfa0beaf358dd9e.

jchristgit commented:
@ilyam8 Any idea about the healthchecks not displaying in the alarms page on the web dashboard?

ilyam8 commented:
@jchristgit 
it seems you need to add them manually to the
https://github.com/netdata/netdata/blob/master/health/Makefile.am

ilyam8 commented:
@jchristgit i have no `riak kv`, so i cant properly test the collector. Please make sure it works on py2/py3, it survives riak restart and there is no problems with alarms (valid calculations, etc).

ilyam8 commented:
Very nice work @jchristgit 	:+1: LGTM

jchristgit commented:
@ilyam8 Thanks for the feedback.

As for the healthchecks, it seems that the Makefile entry was also missing for proper installation - I dropped the health configuration file as-is into `/etc/netdata/health.d/riakkv.conf` on the machine I'm testing with. However, it seems like that's not being processed in the first place, which is my issue. I will check the other points you mentioned.

ilyam8 commented:
>  I dropped the health configuration file as-is into /etc/netdata/health.d/riakkv.conf

make sure the file is readable by `netdata` user

___

and it is not needed, installer will add the alarm, just intall your branch

jchristgit commented:
Tested surviving Riak restarts and Python 2 compatability, both work.

The configuration file is readable for netdata - no error in the logfile either (disabled log flood protection to make sure I'm not missing something).

Maybe it'd be better to leave out the health configuration setup from this PR until I have it figured out? I can open up a separate PR once I get it to work.

ilyam8 commented:
>  I just figured this out - I need to reference the context, not the context + the metric. I'll update the plugin accordingly

this is true for templates, yes

jchristgit commented:
I've simplified two metric names, fixed the existing health checks and added two more health checks. My colleague does not have any other ideas for health checks or metrics to collect either, so this would be done from my side.

ilyam8 commented:
i installed the PR 

> [ilyam-pc plugins.d]# ./python.d.plugin riakkv
2019-06-18 15:51:31: python.d INFO: plugin[main] : using python v3
2019-06-18 15:51:31: python.d ERROR: plugin[main] : unknown modules : ['riakkv']
DISABLE

please add the plugin to the
https://github.com/netdata/netdata/blob/83d6968b19101accc4f424aa8a2a49487a8d8b25/collectors/python.d.plugin/Makefile.am

jchristgit commented:
Added to the Makefile. Also fixed an issue related to looking up the current value on latency healthchecks instead of the three-minute average.

ilyam8 commented:
i installed the branch, lgtm
